iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/storable_function.inl

```diff
@@ -0,0 +1,343 @@
+        // (this cannot happen in the static_storage case)
+    }
+
+    cxx::Ensures(!empty());
+
+    // We detect the problem at compile time or store nothing when memory is exhausted.
+    // Note that we have no other choice, it is used in the ctor and we cannot throw
+    // the object will be valid but not callable (operator bool returns false).
+
+    // If used with StorageType = static_storage we always detect this at compile time.
+}
+
+template <typename S, typename ReturnType, typename... Args>
+template <typename CallableType>
+void storable_function<S, signature<ReturnType, Args...>>::copy(const storable_function& src,
+                                                                storable_function& dest) noexcept
+{
+    if (src.empty())
+    {
+        // nothing to do, destroy and setting m_invoker to nullptr are performed before the type specific copy
+        // operation
+        return;
+    }
+
+    auto ptr = dest.m_storage.template allocate<CallableType>();
+
+    if (ptr)
+    {
+        auto obj = reinterpret_cast<CallableType*>(src.m_callable);
+        ptr = new (ptr) CallableType(*obj);
+        dest.m_callable = ptr;
+        dest.m_invoker = src.m_invoker;
+    }
+    else
+    {
+        std::cerr << "storable_function: no memory to store copy at destination\n";
+        // no memory avilable in source, assignment could not be performed
+        // (this cannot happen in the static_storage case)
+    }
+
+    cxx::Ensures(!dest.empty());
+}
+
+template <typename S, typename ReturnType, typename... Args>
+template <typename CallableType>
+void storable_function<S, signature<ReturnType, Args...>>::move(storable_function& src,
+                                                                storable_function& dest) noexcept
+{
+    if (src.empty())
+    {
+        // nothing to do, destroy and setting m_invoker to nullptr are performed before the type specific move
+        // operation
+        return;
+    }
+
+    auto ptr = dest.m_storage.template allocate<CallableType>();
+    if (ptr)
+    {
+        auto obj = reinterpret_cast<CallableType*>(src.m_callable);
+        ptr = new (ptr) CallableType(std::move(*obj));
+        dest.m_callable = ptr;
+        dest.m_invoker = src.m_invoker;
+        src.m_operations.destroy(src);
+        src.m_callable = nullptr;
+        src.m_invoker = nullptr;
+    }
+    else
+    {
+        // no memory in source, assignment could not be performed
+        // (this cannot happen in the static_storage case)
+        std::cerr << "storable_function: no memory to store moved object at destination\n";
+    }
+
+    cxx::Ensures(!dest.empty());
+}
+
+template <typename S, typename ReturnType, typename... Args>
+template <typename CallableType>
+void storable_function<S, signature<ReturnType, Args...>>::destroy(storable_function& f) noexcept
+{
+    // TODO: this is not an indication allocate was called anymore (due to fptrs also being stored in m_callable)
```

Comment:
I will solve this in this PR.

Reply:
Looking at this we are safe here since we are always free to call deallocate, even if we did not call allocate (it is like freeing a nullptr, which is a no-op). This is true for the `static_storage` but must be true with other storage types as well (i.e. behave like freeing a `nullptr` in this regard). In other words, there is nothing to do except to remove the todo, no other variable has to be introduced (and this is good since checking lots of flags if not needed should be avoided).